### PR TITLE
Add ResourceConsolidator

### LINKED ISSUE #2103: Upload Work Manager - API refactoring
Discussed this with @anchita-g. One alternative way of handling this is to make the `UploadWorkManager` a little more stateful, similar to the `DownloadWorkManager`. In that case, there would be an API to feed local changes to the `UploadWorkManager`, there will be an API to pull next upload request to be sent to the FHIR server, and there'll be an API to get the sync status - almost like an iterator with a sync progress.

That would make the job of `Uploader` a little simpler - the while loop currently in `UplaoderImpl` would be simplified (because the work is done by the `UploadWorkManager`. This affords the developers a little more control - but admittedly we're not entirely sure if the additional contronl is urgently needed by the developers.

But that would make the design more alinged with the download workflow which @omarismail94 worked on.

FYI @aditya-07 as well.

_Originally posted by @jingtang10 in https://github.com/google/android-fhir/pull/2022#discussion_r1277357351_

### PR DESCRIPTION
**Pull request by omarismail94**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #2103

**Description**
Clear and concise code change description. 

The resource consolidator would handle any post-processing that needs to be made to the database. Examples of this would be, updating the lastUpdated timestamp field, or deleting the local change from the database, or updating the resource IDs and payloads to correspond with the server’s feedback.

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: (Bug fix | Feature | Documentation | Testing | Code health | Builds | Releases | Other)

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: engine/src/main/java/com/google/android/fhir/impl/FhirEngineImpl.kt [hunk 1]
```diff
@@ -123,89 +123,17 @@ internal class FhirEngineImpl(private val database: Database, private val contex
...
-    upload: suspend (List<LocalChange>) -> Flow<Pair<LocalChangeToken, Resource>>
+    uploadMode: UploadMode,
+    upload: suspend (List<LocalChange>, ResourceConsolidator) -> Unit,
   ) {
+    val resourceConsolidator =
+      when (uploadMode.resourceConsolidatorStrategy) {
+        ResourceConsolidatorStrategy.Default -> DefaultResourceConsolidator(database)
+        ResourceConsolidatorStrategy.IDUpdater -> error("Not yet implemented")
+      }
     val localChanges = database.getAllLocalChanges()
```
Comment (anchita-g) [inline engine/src/main/java/com/google/android/fhir/impl/FhirEngineImpl.kt:134]: We would have to modify the fetcher as well based on the upload mode

File: engine/src/main/java/com/google/android/fhir/sync/upload/UploaderImpl.kt [hunk 2]
```diff
@@ -43,7 +43,10 @@ internal class UploaderImpl(
   private val uploadWorkManager: UploadWorkManager,
 ) : Uploader {
 
-  override suspend fun upload(localChanges: List<LocalChange>): Flow<UploadState> = flow {
+  override suspend fun upload(
+    localChanges: List<LocalChange>,
```
Comment (anchita-g) [inline engine/src/main/java/com/google/android/fhir/sync/upload/UploaderImpl.kt:47]: Would we not need `UploadMode` here to decide how to create patches of request and request builder.

Comment (omarismail94) [reply engine/src/main/java/com/google/android/fhir/sync/upload/UploaderImpl.kt:47]: I didn't add the patch and request builder as they have yet to be created. In the future, the upload function would look like this:

```
override suspend fun upload(localChangeFetcher: LocalChangeFetcher, patchGenerator: PatchGenerator, requestBuilder: RequestBuilder, resourceConsolidator: ResourceConsolidator): Flow<UploadState>
```

determining what to pass into the `upload` function based on UploadMode can be done either in the FhirSynchronizer, or in FhirEngineImpl.syncUpload. For the latter, it is for interfaces whose initialisation relies on variables that are only accessible from FhirEngine, such as `Database`

Comment (omarismail94) [reply engine/src/main/java/com/google/android/fhir/sync/upload/UploaderImpl.kt]: Obsolete now as I removed UploadMode from this PR

File: engine/src/main/java/com/google/android/fhir/FhirEngine.kt [hunk 3]
```diff
@@ -54,7 +55,8 @@ interface FhirEngine {
    * api caller should [Flow.collect] it.
    */
   suspend fun syncUpload(
-    upload: (suspend (List<LocalChange>) -> Flow<Pair<LocalChangeToken, Resource>>)
+    consolidatorMode: ConsolidatorMode,
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/FhirEngine.kt:58]: this is unfortuantely glossed over in our previous discussions. do we really need a consolidator mode? or maybe one consolidator is enough? the same consolidator can clear the local changes and update resource id if needed?

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: engine/src/main/java/com/google/android/fhir/sync/upload/ResourceConsolidator.kt [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+      when (mode) {
+        is ConsolidatorMode.Default -> DefaultResourceConsolidator(database)
+        else -> error("$mode does not have an implementation yet.")
+      }
+  }
+}
+
+internal class DefaultResourceConsolidator(private val database: Database) : ResourceConsolidator {
+
+  override suspend fun consolidate(localChangeToken: LocalChangeToken, response: Resource) {
```
Comment (anchita-g) [inline engine/src/main/java/com/google/android/fhir/sync/upload/ResourceConsolidator.kt:39]: Will the information from the `localChangeToken` be enough to perform resource consolidation in more complex cases like ID updation, I was thinking there needs to be some kind of indicator for if the changes were at a per resource level, or were they POST changes, was the request a bundle?

File: engine/src/main/java/com/google/android/fhir/FhirEngine.kt (hunk 3 above)
Comment (omarismail94) [reply engine/src/main/java/com/google/android/fhir/FhirEngine.kt]: I thought about this, and I think one consolidator should be enough. Within the default consolidator, we could have when statements based on the HTTP verb

File: engine/src/main/java/com/google/android/fhir/sync/upload/ResourceConsolidator.kt (hunk 4 above)
Comment (omarismail94) [reply engine/src/main/java/com/google/android/fhir/sync/upload/ResourceConsolidator.kt:32]: I think it should be? If it isn't, it should be able to figure out the consolidation by querying the database (that could come at a performance cost)

File: engine/src/main/java/com/google/android/fhir/sync/upload/ResourceConsolidator.kt [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+package com.google.android.fhir.sync.upload
+
+import com.google.android.fhir.LocalChangeToken
+import com.google.android.fhir.db.Database
+import org.hl7.fhir.r4.model.Bundle
+import org.hl7.fhir.r4.model.Resource
+import org.hl7.fhir.r4.model.ResourceType
+import timber.log.Timber
+
+/** Represents a mechanism to consolidate resources after they are uploaded. */
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/sync/upload/ResourceConsolidator.kt:26]: this kdoc can probably be enriched a bit - why do we have this? what is this class actually gonna do? any edge cases?

**Pull request merged**